### PR TITLE
ARROW-16754: [Java] StructVector's child vectors get unexpectedly reordered after adding duplicated fields

### PR DESCRIPTION
**Pull request by zhztheplayer**

### CODE REVIEW

**Comment by pitrou** [comment]
@lidavidm @davisusanibar Would one of you want to review this?

File: java/vector/src/test/java/org/apache/arrow/vector/TestStructVector.java [hunk 1]
```diff
@@ -171,13 +174,102 @@ public void testAddOrGetComplexChildVectors() {
...
+      initFields.add(Field.nullable("varchar2", MinorType.VARCHAR.getType()));
+      initFields.add(Field.nullable("int2", MinorType.INT.getType()));
+      initFields.add(Field.nullable("varchar3", MinorType.VARCHAR.getType()));
+      initFields.add(Field.nullable("int3", MinorType.INT.getType()));
+
+      // To ensure duplicated field names don't mess up the original field order
+      // in the struct vector
+      initFields.add(Field.nullable("varchar1", MinorType.VARCHAR.getType()));
+      initFields.add(Field.nullable("varchar2", MinorType.VARCHAR.getType()));
+      initFields.add(Field.nullable("varchar3", MinorType.VARCHAR.getType()));
```
Comment (pitrou) [inline java/vector/src/test/java/org/apache/arrow/vector/TestStructVector.java:207]: Should we also test a duplicated field name with a different type?

**Review by lidavidm — COMMENTED** [review]
LGTM overall, with same coment as Antoine. 

Should we also test the behavior with VectorSchemaRoot and unions? Presumably the same problem applied to them too.

**Comment by lidavidm** [comment]
CC @davisusanibar if you'd like to carry this over the finish line

File: java/vector/src/test/java/org/apache/arrow/vector/TestStructVector.java (hunk 1 above)
Comment (zhztheplayer) [reply java/vector/src/test/java/org/apache/arrow/vector/TestStructVector.java:208]: Done

**Comment by zhztheplayer** [comment]
I've addressed the last comment @pitrou @lidavidm please let me know if other changes are needed.

**Review by lidavidm — APPROVED** [review]

**Pull request merged**